### PR TITLE
Resolve columns coming from with queries

### PR DESCRIPTION
**Pull request by collerek**
Missing piece for resolving the subqueries -> resolve also columns coming from with queries.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -3,7 +3,7 @@ name = "sql_metadata"
 version = "2.2.1"
 license="MIT"
 description = "Uses tokenized query returned by python-sqlparse and generates query metadata"
-authors = ["Maciej Brencz <maciej.brencz@gmail.com>", "collerek <collerek@gmail.com>"]
+authors = ["Maciej Brencz <maciej.brencz@gmail.com>", "Radosław Drążkiewicz <collerek@gmail.com>"]
```
Comment (macbre) [inline pyproject.toml:6]: Nice to meet you 🙂

Comment (collerek) [reply pyproject.toml:6]: Haha, nice to meet you too :)

**Comment by collerek** [comment]
Not sure how you want to proceed, wait for this one before new bugs?

File: sql_metadata/parser.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+# pylint: disable=C0302
```
Comment (macbre) [inline sql_metadata/parser.py:1]: > Used when a module has more lines than the limit specified in the max-module-lines option.

Shouldn't we refactor this quite huge module instead of ignoring the linter? Let's do it in a separate PR.

File: sql_metadata/parser.py [hunk 3]
```diff
@@ -562,6 +569,45 @@ def with_names(self) -> List[str]:
         self._with_names = with_names
         return self._with_names
 
+    @property
+    def with_queries(self) -> Dict[str, str]:
+        """
+        Returns with subqueries with names
```
Comment (macbre) [inline sql_metadata/parser.py:575]:
```suggestion
        Returns "WITH" subqueries with names
```

I believe it's easier to read then a sentence with two withs 😉

File: sql_metadata/parser.py (hunk 2 above)
Comment (collerek) [reply sql_metadata/parser.py:1]: Yes that's what I though too - refactor in separate pr.

It again became quite complex as it used to be - it's a "Structured" Query language but it's damned complicated to parse 🤣  

I was thinking about maybe adding `token_type` (or even subclasses and polymorphism) to our token and assigning a type for each token. That way it should speed up the parsing as one token cannot have two functions in particular query (if it's an alias it cannot be a table etc.) so we can skip all tokens that already have a type in all methods. That probably will simplify also some of the checks as we are repeating checks on the same tokens all the time but in different context/functions.

Comment (macbre) [reply sql_metadata/parser.py:1]: Makes sense, let's tackle this one in a separate PR or even discuss them earlier on GitHub.

**Comment by collerek** [comment]
If you need any explanation or further comment let me know.

The one tricky thing was with queries with column aliases like
```
With with_qry (c1, c2) as (select c3, c4 from table)
select with_qry.c1, with_qry.c2 from with_qry 
```

When you want to resolve c1 and c2 they do not appear in with_query, but thay are already included in column_aliases.
So for each with_query I check if this particular query have columns definition and if it does it's skipped from columns, but the position of the token is added to aliases (which will later be added to columns_dict).

So in example above aliases_dict will have `select: [c1, c2]` and it will be resolved based on columns aliases.

If it's with query without columns the search happens inside the with_query normally:
```
With with_qry as (select c3 as c1, c4 as c2 from table)
select with_qry.c1, with_qry.c2 from with_qry 
```

Another tricky part was selecting by wildcard, but in main query using the names of the columns
```
With with_qry as (select * from table)
select with_qry.c1, with_qry.c2 from with_qry 
```
In that case c1 and c2 are resolved to `*` as there is no way to tell from which table (if they are multiple tables like join in with query) the c1 and c2 columns come from.

File: sql_metadata/parser.py [hunk 4]
```diff
@@ -562,6 +569,45 @@ def with_names(self) -> List[str]:
...
+        """
+        Returns "WITH" subqueries with names
+
+        E.g. WITH tableFromWith AS (SELECT * FROM table3)
+             SELECT "xxxxx" FROM database1.tableFromWith alias
+             LEFT JOIN database2.table2 ON ("tt"."ttt"."fff" = "xx"."xxx")
+        will return {"tableFromWith": "SELECT * FROM table3"}
+        """
+        if self._with_queries is not None:
+            return self._with_queries
```
Comment (macbre) [inline sql_metadata/parser.py:583]: I'm wondering if we can introduce a decorator for functions that follow this caching / memoization pattern. Instead of adding this `if` statement we can just annotate the function. But this one is for another PR 🙂

**Review by macbre — APPROVED** [review]

**Comment by macbre** [comment]
@collerek - feel free to merge it 👍🏻

File: sql_metadata/parser.py (hunk 4 above)
Comment (collerek) [reply sql_metadata/parser.py:583]: Yes we can and I agree we should -> can be a part of the refactor with the tokens to simplify again the logic, nice idea :)
